src/hotspot/share/gc/g1/g1VMOperations.cpp

```diff
@@ -50,7 +50,9 @@ bool VM_G1CollectFull::skip_operation() const {
 void VM_G1CollectFull::doit() {
   G1CollectedHeap* g1h = G1CollectedHeap::heap();
   GCCauseSetter x(g1h, _gc_cause);
-  g1h->do_full_collection(false /* clear_all_soft_refs */,
+  bool clear_all_soft_refs = _gc_cause == GCCause::_metadata_GC_clear_soft_refs ||
+                             _gc_cause == GCCause::_wb_full_gc;
```

Comment:
Now that three collectors use this, it might be useful to factor this out as a helper function in `CollectedHeap` and use it.